### PR TITLE
Alerting: Alerts page performance improvements

### PR DESCRIPTION
**Pull request by konrad147**
## Overview

This PR introduces significant performance optimizations for the Alerting Triage page, focusing on reducing CPU overhead during data transformation and visualization rendering. The changes include lazy-loading visualization panels and implementing a more efficient data transformation algorithm.

### Performance Optimizations

1. **Lazy-loading visualization panels** - Uses IntersectionObserver to render expensive visualization components only when they approach the viewport, preventing unnecessary CPU operations for off-screen elements.

2. **Optimized data transformation** - Implements a new single-pass algorithm that is approximately 30x faster than the previous approach when processing large datasets.

3. **Reduced redundant transformations** - Uses `subscribeToState()` instead of `useState()` to transform data only when it actually changes, avoiding 2-3 unnecessary transformation calls per update.

#### Modified Files

- **`public/app/features/alerting/unified/triage/Workbench.tsx`**
  - Added `isLoading` prop support
  - Updated loading state handling

- **`public/app/features/alerting/unified/triage/scene/AlertRuleSummary.tsx`**
  - Split component into `AlertRuleSummaryViz` (contains expensive hooks) and `AlertRuleSummary` (lazy-loading wrapper)
  - Implemented IntersectionObserver with 100px root margin to start loading when element is approaching viewport
  - Added fallback for browsers without IntersectionObserver support

- **`public/app/features/alerting/unified/triage/scene/Workbench.tsx`**
  - Moved data transformation logic to separate file (`dataTransform.ts`)
  - Refactored to use `subscribeToState()` instead of `useState()` for more efficient state management
  - Added `useTransition` hook for non-blocking updates during data transformation
  - Improved loading state tracking

#### New Files

- **`public/app/features/alerting/unified/triage/scene/dataTransform.ts`**
  - New optimized data transformation function `convertToWorkbenchRows()`
  - Single-pass algorithm that builds tree structure efficiently
  - Direct columnar access to field values for better performance
  - Handles grouping, empty values, and alert instance aggregation

- **`public/app/features/alerting/unified/triage/scene/dataTransform.test.ts`**
  - Comprehensive test suite with 913+ lines of tests
  - Covers empty/invalid data handling, flat lists, single/multi-level grouping, empty values, and edge cases

### IntersectionObserver Implementation

- Uses 100px root margin to start loading when element is approaching viewport
- Falls back to immediate rendering for browsers without IntersectionObserver support
- Maintains layout space with placeholder div while not visible

### Data Transformation Algorithm

- **Previous approach**: Multiple passes through data, creating intermediate objects
- **New approach**: Single-pass algorithm that builds tree structure directly
- **Performance**: ~30x faster for large datasets
- Uses direct columnar access to field values for optimal performance

### State Management

- Replaced `useState()` with `subscribeToState()` to avoid unnecessary re-renders
- Uses `useTransition()` for non-blocking updates during expensive transformations
- Only transforms data when actual data changes, not on every state update

### CODE REVIEW

File: public/app/features/alerting/unified/triage/rows/AlertRuleRow.tsx [hunk 1]
```diff
@@ -29,13 +30,13 @@ export const AlertRuleRow = ({
...
+  const handleDrawerOpen = useCallback(() => {
     setIsDrawerOpen(true);
-  };
+  }, []);
 
-  const handleDrawerClose = () => {
+  const handleDrawerClose = useCallback(() => {
     setIsDrawerOpen(false);
-  };
+  }, []);
```
Comment (konrad147) [inline public/app/features/alerting/unified/triage/rows/AlertRuleRow.tsx:39]: To avoid `OpenDrawerIconButton` re-rendering

File: public/app/features/alerting/unified/triage/scene/Workbench.tsx [hunk 2]
```diff
@@ -32,114 +29,52 @@ export function WorkbenchRenderer() {
...
+        if (Array.isArray(value)) {
+          currentGroupByKeys = value.map((value) => String(value));
+        }
+      }
+
+      const { series } = newState.data;
+      // Use transition for non-blocking update
+      startTransition(() => {
+        setRows(convertToWorkbenchRows(series, currentGroupByKeys));
       });
```
Comment (konrad147) [inline public/app/features/alerting/unified/triage/scene/Workbench.tsx:63]: I changed the way we convert time series to table rows. This process can be quite expensive for many rows and it's quite tricky to limit the number of recalculations relying only on `const { data } = runner.useState();` because there are many things that are changing and triggering rerenders (filters, grouping, timerange)
The new approach is more performant, but also more complex. 

Let me now what you think!

File: public/app/features/alerting/unified/triage/scene/Workbench.tsx [hunk 3]
```diff
@@ -32,114 +29,52 @@ export function WorkbenchRenderer() {
...
+      // Get the groupBy from the scene directly to avoid having groupByVariable in the dependency array
+      let currentGroupByKeys: string[] = [];
+      const groupByVariable = sceneGraph.lookupVariable(VARIABLES.groupBy, runner);
+
+      if (groupByVariable && sceneUtils.isGroupByVariable(groupByVariable)) {
+        const value = groupByVariable.getValue();
+        if (Array.isArray(value)) {
+          currentGroupByKeys = value.map((value) => String(value));
+        }
+      }
```
Comment (konrad147) [inline public/app/features/alerting/unified/triage/scene/Workbench.tsx:57]: This code is run asynchronously (it's a callback when called when the new time series data is available) so using `groupByKeys` from line 20 could lead to reruning this effect multiple times and odd issues

File: public/app/features/alerting/unified/triage/scene/dataTransform.test.ts
Comment (konrad147) [inline public/app/features/alerting/unified/triage/scene/dataTransform.test.ts:1]: I had written these tests first, for the old conversion algorithm. Then I swapped the algorithm to ensure it works the same way

File: public/app/features/alerting/unified/triage/scene/dataTransform.ts [hunk 4]
```diff
@@ -0,0 +1,158 @@
+import { DataFrame } from '@grafana/data';
+
+import { AlertRuleRow, EmptyLabelValue, GenericGroupedRow, WorkbenchRow } from '../types';
+
+// Single-pass implementation: 30x faster than previous approach
+// Builds tree structure in one pass through data, avoiding intermediate row objects
+export function convertToWorkbenchRows(series: DataFrame[], groupBy: string[] = []): WorkbenchRow[] {
```
Comment (konrad147) [inline public/app/features/alerting/unified/triage/scene/dataTransform.ts:6]: This algorithm seem to be faster. I run a few benchmarks to compare its performance with the old one and I feel the results are worth the complexity

| Test | Old (ops/sec) | Old (avg time) | New (ops/sec) | New (avg time) | Speedup |
|------|---------------|----------------|---------------|----------------|---------|
| No grouping | 11.26 | 88.790 ms | 384.62 | 2.600 ms | 34.2x |
| Single groupBy (grafana_folder) | 6.55 | 152.710 ms | 59.31 | 16.860 ms | 9.1x |
| Double groupBy (grafana_folder, alertstate) | 6.13 | 163.020 ms | 43.55 | 22.960 ms | 7.1x |

File: public/app/features/alerting/unified/triage/scene/dataTransform.ts [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    !fieldIndex.has('alertname') ||
+    !fieldIndex.has('grafana_folder') ||
+    !fieldIndex.has('grafana_rule_uid') ||
+    !fieldIndex.has('alertstate')
+  ) {
+    return [];
+  }
+
+  // Get required field value arrays (direct columnar access)
+  const alertnameIndex = fieldIndex.get('alertname');
```
Comment (laurenashleigh) [inline public/app/features/alerting/unified/triage/scene/dataTransform.ts:31]: Could these index values ever be 0? If so these will be falsey so may have to explicitly check if they're undefined?

File: public/app/features/alerting/unified/triage/rows/OpenDrawerIconButton.tsx [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  onClick,
+  ['aria-label']: ariaLabel,
+}: OpenDrawerIconButtonProps) {
+  const styles = useStyles2(getStyles);
+  return <IconButton className={styles.iconButton} name="web-section-alt" aria-label={ariaLabel} onClick={onClick} />;
+});
+
+const getStyles = (theme: GrafanaTheme2) => ({
+  iconButton: css({
+    transform: 'rotate(180deg)',
```
Comment (laurenashleigh) [inline public/app/features/alerting/unified/triage/rows/OpenDrawerIconButton.tsx:22]: Could we avoid calling useStyles on every render by using inline styling as the css is so simple?

Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/triage/rows/OpenDrawerIconButton.tsx:22]: I would keep `useStyles2` for consistency with the codebase patterns and easier theme integration, and I think the performance impact should be negligible here.

File: public/app/features/alerting/unified/triage/scene/dataTransform.ts (hunk 5 above)
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/triage/scene/dataTransform.ts:31]: this is a good point Lauren, but I tink in this case it's not possible, right?

File: public/app/features/alerting/unified/triage/rows/OpenDrawerIconButton.tsx (hunk 6 above)
Comment (konrad147) [reply public/app/features/alerting/unified/triage/rows/OpenDrawerIconButton.tsx:22]: In general using `className` should be faster than inline styles (inline styles needs to be recalculated all the time) - that's why I did this change
Also `useStyles2` is memoized and well optimized so it shouldn't be an issue

File: public/app/features/alerting/unified/triage/scene/dataTransform.ts (hunk 5 above)
Comment (konrad147) [reply public/app/features/alerting/unified/triage/scene/dataTransform.ts:31]: There is already validation above (and below 😅) which ensures the index exists

File: public/app/features/alerting/unified/triage/rows/AlertRuleRow.tsx
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/triage/rows/AlertRuleRow.tsx:23]: what about memoizing this `AlertRuleRow`?

File: public/app/features/alerting/unified/triage/scene/AlertRuleSummary.tsx [hunk 7]
```diff
@@ -91,3 +97,57 @@ export function AlertRuleSummary({ ruleUID }: { ruleUID: string }) {
...
+
+    observer.observe(container);
+
+    return () => {
+      observer.disconnect();
+    };
+  }, [ruleUID]);
+
+  return (
+    <div ref={containerRef} style={{ width: '100%', height: '100%' }}>
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/triage/scene/AlertRuleSummary.tsx]: nit: className + useStyles2 here?

File: public/app/features/alerting/unified/triage/scene/AlertRuleSummary.tsx [hunk 8]
```diff
@@ -91,3 +97,57 @@ export function AlertRuleSummary({ ruleUID }: { ruleUID: string }) {
...
+        rootMargin: '100px', // Start loading when element is 100px away from viewport
+      }
+    );
+
+    observer.observe(container);
+
+    return () => {
+      observer.disconnect();
+    };
+  }, [ruleUID]);
```
Comment (soniaAguilarPeiron) [inline public/app/features/alerting/unified/triage/scene/AlertRuleSummary.tsx]: why do we have this `ruleUID` in the dependency array? I think we should keep it empty, as the observer doesn't depend on ruleUID , it only tracks viewport visibility, so it should be created once on mount

File: public/app/features/alerting/unified/triage/scene/Workbench.tsx (hunk 2 above)
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/triage/scene/Workbench.tsx:63]: I believe this solution it's appropriately complex for the problem 👌 . The subscription pattern + transition is more lines of code, but I'm not sure if we can achieve an improvement like this in a more simple way.
great job!!!!

File: public/app/features/alerting/unified/triage/rows/AlertRuleRow.tsx
Comment (konrad147) [reply public/app/features/alerting/unified/triage/rows/AlertRuleRow.tsx:23]: I'm not sure. There are a few similar components: `GroupRow, FolderGroupRow, InstanceRow` but I didn't have time to verify if memoizing them could bring benefits

File: public/app/features/alerting/unified/triage/scene/AlertRuleSummary.tsx (hunk 8 above)
Comment (konrad147) [reply public/app/features/alerting/unified/triage/scene/AlertRuleSummary.tsx]: Good catch! I guess it's a leftover from some other code version. I'll remove it

**Review by laurenashleigh — APPROVED** [review]
LGTM!

**Review by soniaAguilarPeiron — APPROVED** [review]
Great job!
